core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java

```diff
@@ -213,10 +214,12 @@ Collection<Text> listSplits(String tableName, int maxSplits)
    * @param ranges The input ranges that should be mapped to tablet servers and tablets.
    *
    * @throws TableOfflineException if the table is offline or goes offline during the operation
+   * @throws InvalidTabletHostingRequestException if tablet hosting is requested for table with
+   *         NEVER goal
    * @since 1.8.0
    */
-  Locations locate(String tableName, Collection<Range> ranges)
-      throws AccumuloException, AccumuloSecurityException, TableNotFoundException;
```

Comment:
Not sure about binary compatibility, but I think adding or removing checked exceptions breaks source compatibility.  So we usually avoid changing them.  Could make this an unchecked exception and declare it in the javadoc.

Reply:
I didn't think we would care about breaking compatibility since this is the 4.0 branch and for elasticity. Seems like for a major version it would be fine to be incompatible with a new API in this case but I could be wrong.

One plus is if it is a runtime exception you could re-work this and avoid having to add the exception everywhere to rethrow it. Another option to avoid having to add it to the throws line everywhere is to update the new exception to  just extend AccumuloException and still handle it separately. However, this would not follow the typical pattern as all the existing exceptions just seem to extend Exception or RuntimeException.

Reply:
If the method is only changing the exceptions - it may not be possible to mark the method as deprecated and provide the "new" method - if the only difference is the exceptions in the method signature.

Reply:
I used the checked exception because it caused things to not-compile and break in the IDE. That allowed me to figure out how/where to handle the exception. If I were to throw an AccumuloException or RuntimeException from the ClientTabletCacheImpl.requestTabletHosting method, then it would likely have been swallowed and in some cases not handled correctly. Regarding the new exception, unlike a TableNotFoundException, the client could retry the operation after removing the ranges that are set to NEVER.

Reply:
> I didn't think we would care about breaking compatibility since this is the 4.0 branch and for elasticity.

We don't usually break the compatibility of method in place.  We usually deprecate them and introduce new methods.  This allows code to keep compiling while devs work through removing deprecated code.  As @EdColeman  mentioned when you only want to change the checked exceptions on a method, Java provides no good way to slowly evolve this that I know of.

Reply:
> Regarding the new exception, unlike a TableNotFoundException, the client could retry the operation after removing the ranges that are set to NEVER.

That is a good thing .  Its nice to have things that people can reliably write automation against for specific error conditions rather than parsing the exception message.

Reply:
My take-away from this thread is that I have nothing to do here. Correct?

Reply:
If we make InvalidTabletHostingRequestException a RunTimeException instead of an Exception we can avoid an incompatible change.

Reply:
InvalidTabletHostingRequestException could also extend AccumuloException and void a breaking change.  This avoids the breaking change while still passing more specific information that can easily be automatically handled.

Reply:
> If we make InvalidTabletHostingRequestException a RunTimeException instead of an Exception we can avoid an incompatible change.

If we make it a RuntimeException, then existing code will compile successfully and users may not realize that they can catch the exception, modify their parameters, and retry the operation to get a successful response.

> InvalidTabletHostingRequestException could also extend AccumuloException and void a breaking change. This avoids the breaking change while still passing more specific information that can easily be automatically handled.

This has the same problem as the RuntimeException. In addition, there can be places in our code where we need to handle the InvalidTabletHostingRequestException differently than a generic AccumuloException (ref [this](https://github.com/apache/accumulo/pull/3329/files#diff-220fceec2f35665a23a2111c5d354962d923a7183130529fdacbaa0f3321ec84R155)).


I'm not concerned about breaking compatibilty, as I think the elasticity branch is going to end up being a 4.0.0 release. If adding this Exception to this method is still an issue with a major release, then I would suggest that we deprecate `TableOperations.locate` in `3.0.0` and then rename `locate` to `locateTablets` or something in the `elasticity` branch (which again, would cause the `elasticity` branch to be a 4.0.0 version).

Reply:
I think I have a solution which will not require the API change. The solution is to create `ClientTabletCache.binHostedRanges` which calls `findTablets` with `LocationNeed.NOT_REQUIRED` and catches the Exception. it occurred to me after writing my last comment, that merely finding tablet locations should not require hosting.

Reply:
This is the API method that assumes all tablets have a location.  https://github.com/apache/accumulo/blob/1fd1284afc270366d4dfadb79ee285fab5b060c9/core/src/main/java/org/apache/accumulo/core/client/admin/Locations.java#L53

Reply:
Modified TableOperationsImpl.locate to call findTablets instead of binRanges, in 6744e61

Reply:
I looked at the changes in 6744e61 and it changes the behavior behind the API.  The API used to return locations for every tablet in the given range.  Now it returns a subset of the tablets in the range that happen to have a location.  For anything using this API the behavior change could cause it to break in weird ways.    

We can not leave the API as is because it could cause tablets to be unexpectedly hosted.  The API just does not seem viable with the on demand changes, I don't see a way to make it do anything sensible.

I can take a stab at deprecating this in 3.x, then we can delete the API in the elasticity branch.   Let me know if you see any issues with that.

Reply:
This API made me wonder about map reduce and I ended up opening #3341

Reply:
> I looked at the changes in [6744e61](https://github.com/apache/accumulo/commit/6744e611c514ecebe41c37f0ea398ef2ee1e8219) and it changes the behavior behind the API. The API used to return locations for every tablet in the given range. Now it returns a subset of the tablets in the range that happen to have a location. For anything using this API the behavior change could cause it to break in weird ways.
> 
> We can not leave the API as is because it could cause tablets to be unexpectedly hosted. The API just does not seem viable with the on demand changes, I don't see a way to make it do anything sensible.
> 
> I can take a stab at deprecating this in 3.x, then we can delete the API in the elasticity branch. Let me know if you see any issues with that.

Yeah, I was trying to just locate the tablets that are hosted, not cause others to be hosted just by trying to get the location. If you think that is the best course of action, I have no issue with that.

Reply:
I had one more idea, maybe we can just make this only w/ a hosting goal of always.  Going to pull your branch and add that check and see what it looks like.

Reply:
I made a comment with an alternative approach to deprecation.

Reply:
I applied your patch so that I could better see the differences. For the tablets that are returned, I don't think all of them have a goal of ALWAYS as the one's returned in the initial call to `locator.findTablets` could be currently hosted, but with a goal of ONDEMAND. The retry code checks that the tablets have a goal of ALWAYS. This change also clears the TabletLocator, so any other concurrent operation may incur a latency penalty when it tries to scan / write to a Tablet. Since this change introduces a call to Ample to check that the `locationless` are ALWAYS hosted, I wonder if `TableOperations.locate` should just call Ample and return the `Locations` for ALWAYS hosted Tablets for the requested table and not clear/use the TabletLocator.

Reply:
I like adding the goal to the cache.